### PR TITLE
perf tweak to run OpenFileOnly analyzers only for open files even if …

### PR DESCRIPTION
**Pull request by heejaechang**
…full solution analysis is on but still run for fix all for closed files.

also made add using test to exercise OOP code properly

...

we have OpenFileOnly flag that we use to indicate to the engine that we want analyzers to run only on open files.

we have this flag since we have certain slow analyzers that we want to turn on by default rather than just disable it and let users to enable if they want and pay perf cost. example of such analyzer is SimplifyTypeNameAnalyzer. there can be more but that is one I know.

this PR doesn't discuss whether that flag is needed or not or need to be deleted. 

if some analyzer behaves differently due to this bug fix than we just discovered new bug on analyzers.

@sharwell seems looked through OpenFileOnly analyzers so if he found buggy analyzers we can add that to the issue I opened so that we can fix those analyzers by returning false for open files if one believes it must run for all files.

also, why some analyzers are open file only or not is not this PR's concern. I believe someone made the decision in some design meeting and set it following the decision. if one wants to question all those decisions, I am fine with that, but that is not about this PR.

### CODE REVIEW

File: src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.InProcOrRemoteHostAnalyzerRunner.cs [hunk 1]
```diff
@@ -66,8 +60,9 @@ public InProcOrRemoteHostAnalyzerRunner(DiagnosticAnalyzerService owner, Abstrac
                     return await AnalyzeInProcAsync(analyzerDriver, project, cancellationToken).ConfigureAwait(false);
                 }
 
-                // due to OpenFileOnly analyzer, we need to run inproc as well for such analyzers
-                var inProcResultTask = AnalyzeInProcAsync(CreateAnalyzerDriver(analyzerDriver, a => a.IsOpenFileOnly(project.Solution.Workspace)), project, remoteHostClient, cancellationToken);
+                // due to OpenFileOnly analyzer, we need to run inproc as well for such analyzers for fix all
+                // otherwise, we don't need to run open file only analyzers for closed files even if full solution analysis is on (perf improvement)
```
Comment (heejaechang) [inline src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.InProcOrRemoteHostAnalyzerRunner.cs]: this only add OpenFileOnly analyzer if forcedAnalysis is true.

File: src/VisualStudio/Core/Test.Next/Services/VisualStudioDiagnosticAnalyzerExecutorTests.cs [hunk 2]
```diff
@@ -189,8 +189,11 @@ void Method()
                         analyzerReference.GetAnalyzers(project.Language).Where(a => a.GetType() == analyzerType).ToImmutableArray(),
                         new WorkspaceAnalyzerOptions(project.AnalyzerOptions, project.Solution.Options, project.Solution));
 
+                // no result for open file only analyzer unless forced
```
Comment (heejaechang) [inline src/VisualStudio/Core/Test.Next/Services/VisualStudioDiagnosticAnalyzerExecutorTests.cs]: this is a test that shows the change.

**Comment by heejaechang** [comment]
I left comments for 2 places where project actually affected and test that touch the change.

all others are test infrastructure only changes that I cleaned up while touching code around analyzers.

some code fix test didn't exercise OOP code and now it should through remote host mock.

**Comment by heejaechang** [comment]
tagging @mavasani can you take a look? this is targetting 16.2 unless @jinujoseph want this for 16.1

**Comment by heejaechang** [comment]
@mavasani by the way, your unused parameter thing works amazing. next step might be us adding refactoring to remove unused parameter which also follows all callers and propagate all the cleanups.

right now, I did it manually but still that visible indication of unused parameter made cleaning up so much easy especially when following all callers and propagating cleanups.

**Comment by sharwell** [comment]
I would strongly prefer to move in a direction where diagnostics are no longer marked as open file only. Any current coding issue that made open file only not work as originally intended makes it easier to determine that eliminating the designation would not cause problems.

Was this change fixing a measured performance problem or just cleaning up a limitation you found while reviewing code?

**Comment by heejaechang** [comment]
@sharwell things I found while looking around code. and i confirmed that we run less analyzer in proc when FAS is on with this change.

about removing OpenFileOnly analyzers. I am not against we removing OpenFileOnly analyzers and when we do, we can remove all code related to OpenFileOnly.

but until then, improving perf around OpenFileOnly seems not bad thing to do if we can. no reason for users to suffer until we do.

**Comment by heejaechang** [comment]
by the way, the reason I start to look around code was beacuse manish told me there is a feedback saying VS feels slower than before when full solution analysis is on. I dont know whether that is due to this or not. but surely this will improve perf of VS when full solution analysis is on.

**Comment by heejaechang** [comment]
oops. wrong butten clicked.

File: src/EditorFeatures/TestUtilities/Diagnostics/DiagnosticProviderTestUtilities.cs [hunk 3]
```diff
@@ -36,14 +36,14 @@ public static async Task<IEnumerable<Diagnostic>> GetAllDiagnosticsAsync(Diagnos
 
         public static async Task<IEnumerable<Diagnostic>> GetDocumentDiagnosticsAsync(DiagnosticAnalyzer workspaceAnalyzerOpt, Document document, TextSpan span, Action<Exception, DiagnosticAnalyzer, Diagnostic> onAnalyzerException = null, bool logAnalyzerExceptionAsDiagnostics = false, bool includeSuppressedDiagnostics = false)
         {
-            var testDriver = new TestDiagnosticAnalyzerDriver(document.Project, workspaceAnalyzerOpt, onAnalyzerException, logAnalyzerExceptionAsDiagnostics, includeSuppressedDiagnostics);
```
Comment (mavasani) [inline src/EditorFeatures/TestUtilities/Diagnostics/DiagnosticProviderTestUtilities.cs]: Is `logAnalyzerExceptionAsDiagnostics` no longer used? Should we remove the unused parameter?

Comment (heejaechang) [reply src/EditorFeatures/TestUtilities/Diagnostics/DiagnosticProviderTestUtilities.cs]: will do.

**Comment by CyrusNajmabadi** [comment]
Could the current design be written up?  It's not clear to me now what the implications are for things like FAS/OOP/OpenFiles/etc.  What is the expectation between what is run and where, and how these options change things.

Thanks!

**Comment by heejaechang** [comment]
those are written in the code but I guess they are separated out in multiple places. if one can point me where one want these kind of things to be written. I can write it there.

**Comment by genlu** [comment]
@heejaechang the docs directory would be a good place for it:
https://github.com/dotnet/roslyn/tree/master/docs

**Comment by heejaechang** [comment]
@genlu cool. I will put some doc there. but in separate PR.

**Comment by CyrusNajmabadi** [comment]
>I will put some doc there. but in separate PR.

That's fine.  But can we at least get a brief explanation here?  I can't figure out of this code is "right" or "wrong" since i don't know what the designs/expectations are here.

**Comment by heejaechang** [comment]
OpenFileOnly analyzer shouldn't run for FSA. the fact it ran before is just a bug. it should only run when forced (fix all)

Workspace analyzer or host analyzer (vsix) whatever way we call it is not part of Roslyn solution. so someone (usually host) must have sure it is sync to OOP for them to available in OOP.

existing tests did 2 thing wrong.

1. it didn't connect the analyzer to either solution or host so it never eixst in OOP.
2. we can not sync some random code to OOP. for analyzer, we need analyzer reference so that we can load analyzer from it in OOP. but test just passed in some random analzer rather than analyzer reference which can't be synced to OOP.

both are now fixed.

**Comment by CyrusNajmabadi** [comment]
When should somethng be OpenFileOnly?  I see the following are sometimes openfileonly:

1. `use var/explicit-type`.
2. rename-tracking.
3. remove unncessary imports
4. naming styles

Can you explain the reasoning for why these should be openfileonly?

For example, if we make this change hre, and someone has FSA on... we'll stop hearing about naming problems in closed files right?  Same with hearing about var/explicit-types. 

I'm trying to figure out why that would be good.  If the user has FSA on won't that seem like buggy behavior to them?

**Comment by heejaechang** [comment]
why we have OpenFileOnly analyzer or not is not this PR's concern or this PR is trying to solve. it already exists and I am not going to change it. 

if existing analyzers you listed is using OpenFileOnly API wrongly please open bugs to fix those analyzers. OpenFileOnly means it runs only for open files.

also, this PR is not making those analyzers OpenFileOnly or not. you should ask that question to author s of that analyzers.

**Comment by heejaechang** [comment]
and doc seems clearly state that OpenFileOnly returning true means it only runs for open files
http://source.roslyn.io/#Microsoft.CodeAnalysis.Features/Diagnostics/IBuiltInAnalyzer.cs,28

**Comment by CyrusNajmabadi** [comment]
>why we have OpenFileOnly analyzer or not is not this PR's concern 

It is this PR's concern, because this PR is about to break user functionality.  You can't just break functionality and punt on that.  Either we need to decide:

1. the breakage is ok.  but we'll make that decision that affirm it before proceeding.
2. the breakage is not ok, and it needs ot be fixed as part of this PR.
3. the breakage is not ok, but another PR can fix it.

But that has to be decided on.  Note: it was only because i actually took the time to read this PR and investigate it that i even found that this would be a problem.  We should be doing due dilegence here and not just blindly making changes that degrade the experience without even realizing it.

Thsi PR could at least have said: this will change the behavior for the following features, and here's how it will impact the user.  That was not done, leading to there being incomplete information by which to review the PR.

I stand by my assessment.  This user facing side of htis PR needs to be signed off on in terms of whether or not the behavior change is acceptable.

Pulling in @sharwell @mavasani for their thoughts.   Personally, i do not think it's ok that naming-styles would not be checked for closed files even if FSA is on.  We need to make an informed decision about how we deal with that potential change in behavior.

**Comment by CyrusNajmabadi** [comment]
Note: this also indicates a test hole as no tests caught this significant change in behavior.  Had we had a test, we would have had to decide what to do about this breakage.  The lack of a test doesn't change the need to make that decision when we catch regressions through other means (like a thorough review of a PR).

**Comment by heejaechang** [comment]
like I said, if you think those analyzers need to run on closed files, change those analyzers not to say true on OpenFileOnly. rather than asking engine not to ignore the fact that it said OpenFileOnly true and run it for closed file regardless.

and this PR doesn't concern whether why those analyzer choose to return OpenFileOnly true. it only care the fact it right now says true according to you. (I didnt check whether they actually returned true or not)

**Comment by CyrusNajmabadi** [comment]
>like I said, if you think those analyzers need to run on closed files, change those analyzers not to say true on OpenFileOnly.

This is your PR.  It's the one that is changing this user facing behavior.  I'm saying that we need to make an informed choice on what to do about thsi and we can't just punt on things.  I've tagged people i think are appropriate to help make a determination here.

>and this PR doesn't concern whether why those analyzer choose to return OpenFileOnly true. it only care the fact it right now says true according to you. (I didnt check whether they actually returned true or not)

Currently things like the NamingStyleDiagnosticAnalyzerBase have 

```c#
        public override bool OpenFileOnly(Workspace workspace) => true;
```

Your PR is changing behavior here.  It's just part of due-diligence on your part to actually see if there's a change in behavior here that we should be aware of.  If you change behavior you have to actually see if some feature will break because of that.  You can't just say "they were doing things incorrectly, and it's nto my problem".  This is part and parcel of changing core roslyn infrastructure.  If higher levels use your layers, you need to at least see how they're impacted and report on that so people can make informed choices on the best way to proceed.  Otherwise, we can just get regressions, and the issues won't necessarily be found until much later (possibly making it all the way to users).

**Comment by heejaechang** [comment]
no. I dont agree. I am fixing bug. the author of the analyzer made clear choice that these analyzer must only run on open files by return true. and I am fixing the bug we didn't respect the analyzer author's choice.

**Comment by CyrusNajmabadi** [comment]
>the author of the analyzer made clear choice that these analyzer must only run on open files by return true

I'm one of the authors of said analyzer, and i made no such choice.  This is a regression in behavior, and the onus is on your to do due diligence ot determine if that regression is acceptable and desirable.  You don't just get to blindly make changes to core layer and not even check if there is a downstream impact of them.   It's irresponsible behavior.

**Comment by heejaechang** [comment]
I am not sure why you returned OpenFileOnly, true. if you didn't then, someone did. and they made the decision.

and again, open a bug and people will discuss why someone made that decision. and if they want to keep it OpenFileOnly, it will left same, if it is not, it will be changed to false.

this is fixing the engine to follow the contract.

**Comment by CyrusNajmabadi** [comment]
> and I am fixing the bug we didn't respect the analyzer author's choice.

No, that is an asusmption on your part.  For all you know the author made a *mistake*.  Given that OpenFileOnly *only* exists for direct analyzers written by the Roslyn team, the onus is on you to actually figure out what the intent was from the people who wrote this since you work on the very same team as them.  

It should nto be my responsibility to have to grovel through a PR to find issues like thsi, only to find that you're not even willing to take a few minutes to communicate with relevant parties to see if things are ok.  You're making the code change.  Higher layers may have unintentionally taken a dependency on the previous behavior.  So go and actually work with your teammates to determine what hte right thing is to do here.  

Please don't just make a PR that changes behavior and not actually check with people when it changes things.  As i said here, i worked on these features and this may change behavior from our original intent (which was that these were like any sort of normal analyzer).

**Comment by CyrusNajmabadi** [comment]
> I am not sure why you returned OpenFileOnly

primarily because there has been little to no information on this stuff, and we often got it wrong.  That's why i asked this time around for a clear explanation of what this stuff was supposed to do.  It's really not clear to me, and this potentially changes behavior in an unexpected way.  Can you please just work with the respective owners to figure out the right way forward here.

**Comment by heejaechang** [comment]
talk to @jinujoseph 
I am spending more time than I want here.

also, it is funny you are saying things like above. since I told exact same thing on sqlite that some features using it for a state so that it must guarantee if someone saves data to it, it should guarantee read must succeed. and I think you said, since it is not written on contract and you are following the contract it is features' fault that it assume something on implementation. and it is fine for engine to break it and feature must fix it since it is a feature fault.

now we are like completely opposite stand. anyway. I will check in once integration pass.

**Comment by CyrusNajmabadi** [comment]
> also, it is funny you are saying things like above. since I told exact same thing on sqlite that some features using it for a state so that it must guarantee if someone saves data to it, it should guarantee read must succeed. 

1. It provides that guarantee to the same degree as the previous system.  i.e. unless the db got deleted, both systems will succeed on the read.

> and you are following the contract it is features' fault that it assume something on implementation.

I never said this wasn't an issue where the feature wasn't at fault.  The features *may* be at fault.  But that doesn't mean we shouldn't fix it.  I don't see why that is contentious.  If this leads to a break in behavior for the user we should at least be *aware* of it.  I'm not saying you can't make the PR cahne.  I'm saying we should do *due diligence*.

**Comment by CyrusNajmabadi** [comment]
>also, it is funny you are saying things like above. since I told exact same thing on sqlite 

For sqlite i tagged you in the PR as the *very first thing i did* because you were an appropriate stakeholder:  https://github.com/dotnet/roslyn/pull/18343#issuecomment-290640148

I really wish you had participated as we could have maybe caught that mistake on my part where i didn't realize the problem i was introducing.  

Note that i did try to do due diligence.  I examined all downstream clients and attempted to understand everything that was going on.  I tried to pull you in on hte PR, but you didn't participate.

**Comment by heejaechang** [comment]
@CyrusNajmabadi let's say I didn't participate the initial PR but not sure later when I raised the concern what the difference is? once code gets checked in, it can't never be changed?

**Comment by CyrusNajmabadi** [comment]
>@CyrusNajmabadi let's say I didn't participate the initial PR but not sure later when I raised the concern what the difference is? once code gets checked in, it can't never be changed?

The question is about roping in stakeholders and making informed choices about downstream breakages.  I attempted to do that with you for the sqlite PR, and i'm asking you to attempt to do the same with those stakeholders impacted by this change.  

I'm not saying that behavior can't change, or that breaks won't happen.  I'm saying that we should be responsible about thsi stuff and do *due diligence* toward mitigating potential regressions (esp. ones that may be customer facing).

It would have taken less time to argue than to pull in appropriate people btw.  We could have potenitally settled this hours ago.  I don't see why this needs to even be a fight.  Just talk to the team.   You work with them.  You know them.  You all have the best interests of customers at heart.

**Comment by heejaechang** [comment]
opened issue https://github.com/dotnet/roslyn/issues/35388 as @jinujoseph asked.

**Comment by heejaechang** [comment]
@sharwell I can't make integration to pass? is there a known issue? all change I made since last time it passed was adding some comments, I dont think the failure is due to my change?

**Comment by heejaechang** [comment]
@jasonmalinowski @sharwell so what is conclusion on this?

**Comment by CyrusNajmabadi** [comment]
> Since the flag is unused and the goal was to remove the separate behavior between internal and external analyzers, we should remove it outright rather than try and "fix" the operation of this.

I'm very much on board with this.

**Comment by CyrusNajmabadi** [comment]
> I really think we need to understand exactly which analyzers this is impacting, if only so if we have bug reports for those we can attempt to connect them back.

Note: i don't think this will be hard.  all i did locally was do a FindRefs on `OpenFileOnly`.  I then looked for any that were different than just `return false;`.  I think there were only like 6-8 analyzers affected.  So it would be easy to literally just enumerate them and check them manually to see if they're impacted.  

It would also just be a good time to look at them and answer: what behavior do we think is right for the user when they change the particular analyzer settings or the FSA setting.  IMO this would be a very small amount of work to do.

**Comment by heejaechang** [comment]
@CyrusNajmabadi if you already did, just write here what you have found. that would be contributing to the discussion.

**Comment by heejaechang** [comment]
@jasonmalinowski I tried to write more in the description as you asked.

**Comment by CyrusNajmabadi** [comment]
> @CyrusNajmabadi if you already did, just write here what you have found. 

Right now my VS is not in a working state.  I attempted to install an update and things are borked.  Otherwise i would :) .  I'm sure you can do the same operation in the meantime, see which analyzers are affected and pull in the appropriate people.  I think i mentioned this either in this thread, or in a chat with Jinu, but i'm fairly certain that between you, @sharwell, @jasonmalinowski and @dpoeschl you could probably find them all and answer the question of what the right user behavior should be in about 5-10 min.

> that would be contributing to the discussion.

Contributions can come in many forms.  I've tried to help by both doing the legwork i've been able to, as well as giving my own insight and feedback on this issue to reach a positive outcome.  Can i do more?  Certainly, but i also have another full time job, and a broken VS and sometimes i can't do it all :)

**Comment by heejaechang** [comment]
@CyrusNajmabadi what I dont understand is I am not sure why you are asking me to do the work you can do as much as I can for the question you have in your head.

let's say I say, ya, I looked OpenFileOnly of analyzers and it looks like there is no big impact I believe users will have.

if you have follow up question, I firmly believe you should do the research and tell me you have these concern with link. and we can discuss. not the other way around.

**Comment by CyrusNajmabadi** [comment]
> @CyrusNajmabadi what I dont understand is I am not sure why you are asking me to do the work 

because you are teh most qualified and more informed on this topic.

**Comment by CyrusNajmabadi** [comment]
> let's say I say, ya, I looked OpenFileOnly of analyzers and it looks like there is no big impact I believe users will have.

If that's waht you and the relevant stakeholders believe, then great.  You all did the necessary due diligence and have made an informed choice together.

**Comment by CyrusNajmabadi** [comment]
> if you have follow up question, I firmly believe you should do the research

I did.  That's what prompted the big brouhaha initially and got other people involved to weigh in on this topic.  I went and pulled down your change and discovered several things of concerns.  I pointed them out and gave advice on what could be done about it.  You're an experience engineer who is used to working with your team.  So i trust you can take the appropriate steps with them toward a positive resolution for customers.

**Comment by heejaechang** [comment]
then, just tell me what analyzer you have concern with. then I will take a look and let you know.

**Comment by CyrusNajmabadi** [comment]
> then, just tell me what analyzer you have concern with. 

I am concerned with all analyzers that implement OpenFileOnly and don't pass 'false' directly back.  I cannot give you a list of that directly as i do not have a functioning VS.   I trust that this would be pretty simple for you or anyone else on Roslyn to do.  Originally it took me about 30 seconds in VS.  So i don't think it's asking much for you to put in that much effort to help out the downstream clients of your API.

> then I will take a look and let you know.

i would prefer you and the relevant owner of that feature discuss what's going on and come to a conclusion on what they want the right user behavior to be.  As i mentioned 2-3 times before, jsut having a quick chat with Sam/Jason/David would likely lead to a very quick resolution on where to take things.  It would probably only be a few minutes.  I'm sure they'd be happy to work with you on this.  If there are open questions they think they could use my help on, i'm happy to provide any insight or guidance i can.  However, this seemed like something that all your collective expertise woudl be more than able to handle efficiently :)

**Comment by heejaechang** [comment]
@CyrusNajmabadi well, we had 2 hours discussion and Sam said the impact would be the suggestion for certain analyzers not showing up when full solution analysis is on anymore. I think Sam said, there are 2 analyzers. 

but I didn't check since I believe Sam won't just come up the number and say it. but that is all I know. if you need more detail you can ask Sam.

**Comment by CyrusNajmabadi** [comment]
>  if you need more detail you can ask Sam.

I'm happy to let you and Sam hash it out to determine the best course of action forward here.  From some of the issues i've been tracking it sounds like one of hte beliefs is that we should just remove OpenFileOnly from this internal interface.  I would be entirely supportive of that.

**Comment by mavasani** [comment]
@CyrusNajmabadi we (@sharwell @jasonmalinowski @heejaechang @jinujoseph and me) had a meeting yesterday to discuss the concerns and the path forward in this area. We identified the following as the primary concerns/open issues:

1. Add more context on the changes implemented in the PR: Heejae has added more details in the PR description. Additionally, we have got a very number of VS feedback requests recently around live code analysis hogging the CPU (see [here](https://developercommunity.visualstudio.com/content/problem/532880/live-code-analysis-slow.html), [here](https://developercommunity.visualstudio.com/content/problem/543977/live-code-analysis-slow-1.html) and [here](https://developercommunity.visualstudio.com/content/problem/533024/vs2019-code-analysis-hogs-the-cpu-and-never-stops.html)). These requests still do not have performance traces for us to root cause if these are the same issues or not. However, these led to Heejae investigating the background analysis code and find this OpenFileOnly bug in the engine, which can be a likely cause of these performance issues.
2. Perceived end user functionality changes where prior analyzers which set `OpenFileOnly = true` were still getting executed and reported for closed files when they turn on full solution analysis. Sam has done investigation on the analyzers that are likely affected here and narrowed it down to 2 analyzers: simplify type name analyzer and naming style analyzer. The former reports only hidden severity diagnostics by default and returns false for OpenFileOnly in default configuration (see [here](http://source.roslyn.io/#Microsoft.CodeAnalysis.Features/SimplifyTypeNames/SimplifyTypeNamesDiagnosticAnalyzerBase.cs,70)), so is not a real concern unless the users has explicitly bumped up these diagnostics to non-default severity. Naming style rules do [report info diagnostics](http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/NamingStyles/NamingRule.cs,87c275d55c90d809,references) and would have executed for closed files prior to this PR. None of us know of a definitive reason why this analyzer sets OpenFileOnly to true, but we guessed it was probably done due to us not syncing all the options to OOP due to pending editorconfig work. Given the later has now been done, we should be able to switch OpenFileOnly to false for this analyzer and validate the functional and performance characteristics for the naming analyzer with FSA on. Sam mentioned he had 1-2 issues where this was being requested/tracked and will search for them. If he is unable to find such an issue, we can use https://github.com/dotnet/roslyn/issues/35388 to track this work. Given that the current behavior here of naming rules executing for closed files was unintentional and could be a likely cause of performance issues, we felt it is an acceptable behavior change as it will allow us to verify that functionality before re-enabling it and also let us know if we indeed have customers that want naming violations from closed files being reported.
3. Removing the OpenFileOnly internal hook and even the IBuiltInAnalyzer completely to make our internal analyzers on par with 3rd party analyzers: Everyone unanimously believes these are good changes and the right direction forward. We will ideally like to get rid of these hooks completely, but if we do end up needing one or more of them of optimized analyzer execution in the IDE, we want to make these public APIs. Sam has already made progress on removing OpenFileOnly, and we can potentially get rid of that flag soon. Another option might be to replace these internal APIs with custom tags based approach, which means 3rd party analyzers can also declaratively participate in these IDE-only performance benefits. Reaching the best state here will take more time and this work should be done in parallel, but should not be blocking this PR. 

Given that the current PR is addressing a bug in the currently chosen design in the IDE diagnostic analyzer engine, which may be a cause of bunch of VS feedback performance reports in this area and given the analysis above regarding affected analyzers and scheduled follow-up work, we all felt we should take this PR and follow-up on naming rules running for closed files soon enough, while continually making progress on removal of all internal hooks. Let us know if you have any additional concerns here.

**Comment by CyrusNajmabadi** [comment]
> Given the later has now been done, we should be able to switch OpenFileOnly to false for this analyzer and validate the functional and performance characteristics for the naming analyzer with FSA on. Sam mentioned he had 1-2 issues where this was being requested/tracked and will search for them. If he is unable to find such an issue, we can use #35388 to track this work. Given that the current behavior here of naming rules executing for closed files was unintentional and could be a likely cause of performance issues, we felt it is an acceptable behavior change as it will allow us to verify that functionality before re-enabling it and also let us know if we indeed have customers that want naming violations from closed files being reported.

👍 Thanks for looking into this and working toward making an informed decision on this based on user need and perf concerns.

> Removing the OpenFileOnly internal hook and even the IBuiltInAnalyzer completely to make our internal analyzers on par with 3rd party analyzers: Everyone unanimously believes these are good changes and the right direction forward. We will ideally like to get rid of these hooks completely, but if we do end up needing one or more of them of optimized analyzer execution in the IDE, we want to make these public APIs. Sam has already made progress on removing OpenFileOnly, and we can potentially get rid of that flag soon.

Great!

> Another option might be to replace these internal APIs with custom tags based approach, which means 3rd party analyzers can also declaratively participate in these IDE-only performance benefits. Reaching the best state here will take more time and this work should be done in parallel, but should not be blocking this PR.

Agreed.  We can take small steps at a time until we get to htat end point, but we wouldn't need to block anything on that in the meantime.

> Given that the current PR is addressing a bug in the currently chosen design in the IDE diagnostic analyzer engine, which may be a cause of bunch of VS feedback performance reports in this area and given the analysis above regarding affected analyzers and scheduled follow-up work, we all felt we should take this PR and follow-up on naming rules running for closed files soon enough, while continually making progress on removal of all internal hooks. Let us know if you have any additional concerns here.

Makes perfect sense to me.  Thanks for doing the effort to ensure that the appropriate followup investigations and work will happen.

**Pull request merged**